### PR TITLE
Improve localization sync tool

### PR DESCRIPTION
**Pull request by sftim**
PR #24373 added a handy tool to help localizers check whether pages are in sync

This change adds support for checking multiple paths.
For example, you can now:
```bash
find content/id/docs/tasks/tools -type f -print0 | xargs -r -0 scripts/lsync.sh
```
to list all the files for the bahasa Indonesia localization that are out of date, for originals inside https://k8s.io/docs/tasks/tools/

The exit status if no work is needed changes to `0`. If you listed an invalid path or tried to check any file that isn't present in the localization the exit status is `2`, and if the checks completed OK but a file does need localizing then the exit status is `1`.

/area localization

### CODE REVIEW

**Comment by sftim** [comment]
@tengqm would you like to review this? If you're busy, that's fine.

File: scripts/lsync.sh [hunk 1]
```diff
@@ -3,29 +3,68 @@
...
+  fi
+
+  if ! [ -f $EN_VERSION ] && ! [ -f "$LOCALIZED" ]; then
+    printf "File \"%s\" is missing from upstream (English)\n" "$LOCALIZED" >&2
+    continue
+  fi
 
-git diff $LASTCOMMIT...HEAD $EN_VERSION
+  # Last commit for the localized file
+  LASTCOMMIT="$( git log -n 1 --pretty=format:%h -- "$LOCALIZED" )"
```
Comment (tengqm) [inline scripts/lsync.sh:52]: This line wont work?

File: scripts/lsync.sh [hunk 2]
```diff
@@ -3,29 +3,68 @@
...
-# Try get the English version
-EN_VERSION=`echo $LOCALIZED | sed "s/content\/..\//content\/en\//g"`
-if ! [ -f $EN_VERSION ]; then
-  echo "$EN_VERSION has been removed."
+# Validate paths
+VALID=1
+for ARG in "$@"
+do
+  if printf "%s" "$ARG" | grep -E "^/|^\.\./|^\.\.\$|/\.\./|/\.\.\$"; then
+    printf "%s: Invalid path \"%s\"\n" "$0" "$ARG" >&2
```
Comment (tengqm) [inline scripts/lsync.sh:19]: use 'echo' for consistency?

File: scripts/lsync.sh [hunk 3]
```diff
@@ -3,29 +3,68 @@
...
 fi
 
-# Last commit for the localized file
-LASTCOMMIT=`git log -n 1 --pretty=format:%h -- $LOCALIZED`
+EXITSTATUS=0 # assume success
+for LOCALIZED in "$@"
+do
+
+  # Try get the English version
+  EN_VERSION="$( printf "%s" "$LOCALIZED" | sed "s/content\/..\//content\/en\//g" )"
```
Comment (tengqm) [inline scripts/lsync.sh:32]: this line doesn't work?

File: scripts/lsync.sh (hunk 1 above)
Comment (sftim) [reply scripts/lsync.sh:52]: _It works on my machine!_ :wink:

`$( … )` lets you nest things so inner double quotes are fine (that's different from backticks).

File: scripts/lsync.sh (hunk 2 above)
Comment (sftim) [reply scripts/lsync.sh:19]: I picked `printf` because it's more robust. If needed I can convert the other `echo`s to use `printf` as well.

File: scripts/lsync.sh (hunk 3 above)
Comment (sftim) [reply scripts/lsync.sh:32]: See https://github.com/kubernetes/website/pull/24419#discussion_r501386957

If you're seeing an error, let me know what it looks like.

File: scripts/lsync.sh [hunk 4]
```diff
@@ -3,29 +3,68 @@
...
+  fi
 
-git diff $LASTCOMMIT...HEAD $EN_VERSION
+  # Last commit for the localized file
+  LASTCOMMIT="$( git log -n 1 --pretty=format:%h -- "$LOCALIZED" )"
 
-if [ "$?" -eq 0 ]; then
-  echo "$LOCALIZED is still in sync"
-  exit 3
+  git diff --exit-code "$LASTCOMMIT...HEAD" -- "${EN_VERSION}"
```
Comment (tengqm) [inline scripts/lsync.sh:54]: In general, processing multiple files (or a directory) is a different use case for a script like this. When processing a list of files, I'd expect a summary of which file has been changed. When processing a single file, I'd expect to see the detailed diff, if any.
This is not a blocking factor though.

Comment (sftim) [reply scripts/lsync.sh:54]: > When processing a single file, I'd expect to see the detailed diff, if any.
You will!

There's room to add a `--name-only` or similar option in the future.

**Comment by kbhawkey** [comment]
I am reading through the script changes.
One thing I noted in testing the original script:
This command works for me:
 `./scripts/lsync.sh content/zh/docs/concepts/_index.md`
This command does not:
`cd scripts;` `./lsync.sh content/zh/docs/concepts/_index.md`

I ran the updated script and found the output hard to follow. Still looking.

File: scripts/lsync.sh [hunk 5]
```diff
@@ -3,29 +3,68 @@
...
-# Last commit for the localized file
-LASTCOMMIT=`git log -n 1 --pretty=format:%h -- $LOCALIZED`
+EXITSTATUS=0 # assume success
+for LOCALIZED in "$@"
+do
+
+  # Try get the English version
+  EN_VERSION="$( printf "%s" "$LOCALIZED" | sed "s/content\/..\//content\/en\//g" )"
+  if ! [ -f $EN_VERSION ] && [ -f "$1" ]; then
+    printf "File \"%s\" has been removed\n" "${EN_VERSION}" >&2
```
Comment (kbhawkey) [inline scripts/lsync.sh:34]: This tests if the `en` version exists

File: scripts/lsync.sh [hunk 6]
```diff
@@ -3,29 +3,68 @@
...
+    EXITSTATUS=2
+  fi
+
+  if [ -f $EN_VERSION ] && ! [ -f "$LOCALIZED" ]; then
+    printf "File \"%s\" is missing from the localization\n" "$LOCALIZED" >&2
+    if [ $EXITSTATUS -lt 2 ]; then
+      EXITSTATUS=1
```
Comment (kbhawkey) [inline scripts/lsync.sh:41]: This checks if the `en` version exists and if the localized version exists

File: scripts/lsync.sh [hunk 7]
```diff
@@ -3,29 +3,68 @@
...
+    fi
+    continue
+  fi
+
+  if ! [ -f $EN_VERSION ] && ! [ -f "$LOCALIZED" ]; then
+    printf "File \"%s\" is missing from upstream (English)\n" "$LOCALIZED" >&2
```
Comment (kbhawkey) [inline scripts/lsync.sh:47]: This checks if the `en` and the localization versions are missing

File: scripts/lsync.sh [hunk 8]
```diff
@@ -3,29 +3,68 @@
...
 
-if [ "$?" -eq 0 ]; then
-  echo "$LOCALIZED is still in sync"
-  exit 3
+  git diff --exit-code "$LASTCOMMIT...HEAD" -- "${EN_VERSION}"
+
+  if [ "$?" -eq 0 ]; then
+    printf "File \"%s\" is still in sync\n" "$LOCALIZED" >&2
+  else
+    if [ $EXITSTATUS -lt 2 ]; then
```
Comment (kbhawkey) [inline scripts/lsync.sh:59]: This check seems odd. The status is always < 2?

**Comment by kbhawkey** [comment]
Here are the messages from running:

```
find content/id/docs/tasks/tools -type f -print0 | xargs -r -0 scripts/lsync.sh
```

```
diff --git a/content/en/docs/tasks/tools/install-kubectl.md b/content/en/docs/tasks/tools/install-kubectl.md
...

File "content/en/docs/tasks/tools/kubeadm/_index.md" has been removed
File "content/id/docs/tasks/tools/kubeadm/_index.md" is still in sync
diff --git a/content/en/docs/tasks/tools/_index.md b/content/en/docs/tasks/tools/_index.md
...

File "content/en/docs/tasks/tools/install-minikube.md" has been removed
diff --git a/content/en/docs/tasks/tools/install-minikube.md b/content/en/docs/tasks/tools/install-minikube.md
...

Synchronization update required
```

Question: How can a localization file be in sync when the `en` version is removed?
In this case, does "in sync" mean that the localization file does not exist?

**Comment by sftim** [comment]
Revised again; how's this?

**Comment by kbhawkey** [comment]
👀

**Comment by kbhawkey** [comment]
Hi @sftim . I read through the changes and ran the script providing a directory and also providing a single file.
To recap: Localization teams can use this script to identify any `en` language files that have changed in relation to a localized directory or file. This helps authors create new content and maintain existing content.

I suggest to add more documentation in the script. Describe how to use the script (parameter variations) and the expected output.
@tengqm , Any input on the latest changes? Thanks!

Some observations:
- If you provide an `en` directory, the script lists: a commit and the files in the directory that have changed (since last commit?)
- If you provide an `en` file, the script compares the last two commits?
- If you provide a localization directory, the script lists the related `en` files that have changed, for example:
  
```
./scripts/lsync.sh content/es/docs/concepts/configuration/ > test2.log
Changes since commit d3f374c0d

$ more test2.log 
 content/en/docs/concepts/configuration/secret.md | 10 ++++++++++
 1 file changed, 10 insertions(+)
```
or
```
./scripts/lsync.sh content/es/docs/concepts/
Changes since commit d3f374c0d
 content/en/docs/concepts/architecture/controller.md                 |  2 +-
 content/en/docs/concepts/cluster-administration/system-metrics.md   |  1 -
 content/en/docs/concepts/configuration/secret.md                    | 10 ++++++++++
 content/en/docs/concepts/containers/container-lifecycle-hooks.md    |  2 +-
 content/en/docs/concepts/containers/runtime-class.md                | 17 +++++------------
 content/en/docs/concepts/services-networking/ingress-controllers.md |  3 ++-
 content/en/docs/concepts/services-networking/service.md             |  2 +-
 7 files changed, 20 insertions(+), 17 deletions(-)
```

- An example of testing with a single localized file: `./scripts/lsync.sh content/es/docs/concepts/configuration/secret.md > test.log`
  ```
diff --git a/content/en/docs/concepts/configuration/secret.md b/content/en/docs/concepts/configuration/secret.md
index 1413c05d4..e0512f8b9 100644
--- a/content/en/docs/concepts/configuration/secret.md
+++ b/content/en/docs/concepts/configuration/secret.md
@@ -24,6 +24,16 @@ a password, a token, or a key. Such information might otherwise be put in a
 Pod specification or in an image. Users can create Secrets and the system
 also creates some Secrets.
 
+{{< caution >}}
+Kubernetes Secrets are, by default, stored as unencrypted base64-encoded
+strings. By default they can be retrieved - as plain text - by anyone with API
+access, or anyone with access to Kubernetes' underlying data store, etcd. In
+order to safely use Secrets, we recommend you (at a minimum):
+
+1. [Enable Encryption at Rest](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/) for Secrets.
+2. [Enable RBAC rules that restrict reading and writing the Secret](https://kubernetes.io/docs/reference/access-authn-authz/authorization/). Be aware that secrets ca
n be obtained implicitly by anyone with the permission to create a Pod.
+{{< /caution >}}
+
 <!-- body -->
 
 ## Overview of Secrets
```

And some more tests:
```
./scripts/lsync.sh content/es/docs/concepts2/
File "content/es/docs/concepts2/" is missing from upstream (English)

./scripts/lsync.sh content/en/dos/concepts2/
File "content/en/dos/concepts2/" is missing from upstream (English)

./scripts/lsync.sh 2
File "2" is missing from upstream (English)



```

**Comment by tengqm** [comment]
last time I checked the script, I wasn't sure I can understand what have been improved over existing script.

**Comment by sftim** [comment]
I'll park this.
/close
